server/athenian/api/align/queries/goals.py

```diff
@@ -1,15 +1,135 @@
-from typing import Any
+from functools import partial
+from itertools import groupby
+from operator import itemgetter
+from typing import Any, Iterable, Mapping
 
 from ariadne import ObjectType
 from graphql import GraphQLResolveInfo
 
+from athenian.api.align.goals.dates import goal_datetimes_to_dates, goal_initial_query_interval
+from athenian.api.align.goals.dbaccess import fetch_team_goals
+from athenian.api.align.goals.templates import TEMPLATES_COLLECTION
+from athenian.api.align.models import GoalTree, GoalValue, MetricValue, TeamGoalTree, TeamTree
+from athenian.api.align.queries.metrics import calculate_team_metrics
+from athenian.api.align.queries.teams import build_team_tree_from_rows
+from athenian.api.async_utils import gather
+from athenian.api.db import Row
+from athenian.api.internal.account import get_metadata_account_ids
+from athenian.api.internal.team import fetch_teams_recursively
+from athenian.api.internal.with_ import flatten_teams
+from athenian.api.models.state.models import Goal, Team, TeamGoal
 from athenian.api.tracing import sentry_span
+from athenian.api.typing_utils import dataclass
 
 query = ObjectType("Query")
 
 
 @query.field("goals")
 @sentry_span
-async def resolve_goals(obj: Any, info: GraphQLResolveInfo, **kwargs) -> Any:
-    """Serve goals()."""
-    return []
+async def resolve_goals(
+    obj: Any, info: GraphQLResolveInfo, accountId: int, teamId: int, **kwargs,
+) -> Any:
+    """Serve goals() query."""
+    team_rows, meta_ids = await gather(
+        fetch_teams_recursively(
+            accountId,
+            info.context.sdb,
+            select_entities=(Team.id, Team.name, Team.members, Team.parent_id),
+            # teamId 0 means to implicitly use the single root team
+            root_team_ids=None if teamId == 0 else [teamId],
+        ),
+        get_metadata_account_ids(accountId, info.context.sdb, info.context.cache),
+    )
+
+    team_tree = build_team_tree_from_rows(team_rows, None if teamId == 0 else teamId)
+    team_member_map = flatten_teams(team_rows)
+    team_ids = [row[Team.id.name] for row in team_rows]
+    team_goal_rows = await fetch_team_goals(accountId, team_ids, info.context.sdb)
+
+    res = []
+
+    for _, group_team_goal_rows_iter in groupby(team_goal_rows, itemgetter(Goal.id.name)):
+        # a group for every goal
+        group_team_goal_rows = list(group_team_goal_rows_iter)
+        group_teams = {
+            row[TeamGoal.team_id.name]: team_member_map[row[TeamGoal.team_id.name]]
+            for row in group_team_goal_rows
+        }
+        group_goal_row = group_team_goal_rows[0]  # could be any row
+        metric = TEMPLATES_COLLECTION[group_goal_row[Goal.template_id.name]]["metric"]
+
+        calc_metrics = partial(
+            calculate_team_metrics,
+            metrics=[metric],
```

Comment:
Performance: we should calculate all the metrics at once.

Reply:
Both initial and current. The underlying calculators accept multiple time intervals.

Reply:
`calculate_team_metrics` should be extended to accept several time intervals, and to return the equivalent number of metric value trees.

Reply:
now it's really done